flink-python/pyflink/table/types.py

```diff
@@ -15,172 +15,1803 @@
+                       _merge_type(a.value_type, b.value_type, name='value of map %s' % name))
+    else:
+        return a
+
+
+def _infer_schema_from_data(elements, names=None):
+    """
+    Infers schema from list of Row or tuple.
+
+    :param elements: list of Row or tuple
+    :param names: list of column names
+    :return: :class:`RowType`
+    """
+    if not elements:
+        raise ValueError("can not infer schema from empty data set")
+    schema = reduce(_merge_type, (_infer_schema(row, names) for row in elements))
+    if _has_nulltype(schema):
+        raise ValueError("Some column types cannot be determined after inferring")
+    return schema
+
+
+def _need_converter(data_type):
+    if isinstance(data_type, RowType):
+        return True
+    elif isinstance(data_type, ArrayType):
+        return _need_converter(data_type.element_type)
+    elif isinstance(data_type, MapType):
+        return _need_converter(data_type.key_type) or _need_converter(data_type.value_type)
+    elif isinstance(data_type, NullType):
+        return True
+    else:
+        return False
+
+
+def _create_converter(data_type):
+    """
+    Creates a converter to drop the names of fields in obj.
+    """
+    if not _need_converter(data_type):
+        return lambda x: x
+
+    if isinstance(data_type, ArrayType):
+        conv = _create_converter(data_type.element_type)
+        return lambda row: [conv(v) for v in row]
+
+    elif isinstance(data_type, MapType):
+        kconv = _create_converter(data_type.key_type)
+        vconv = _create_converter(data_type.value_type)
+        return lambda row: dict((kconv(k), vconv(v)) for k, v in row.items())
+
+    elif isinstance(data_type, NullType):
+        return lambda x: None
+
+    elif not isinstance(data_type, RowType):
+        return lambda x: x
+
+    # dataType must be RowType
+    names = [f.name for f in data_type.fields]
+    converters = [_create_converter(f.data_type) for f in data_type.fields]
+    convert_fields = any(_need_converter(f.data_type) for f in data_type.fields)
+
+    def convert_row(obj):
+        if obj is None:
+            return
+
+        if isinstance(obj, (tuple, list)):
+            if convert_fields:
+                return tuple(conv(v) for v, conv in zip(obj, converters))
+            else:
+                return tuple(obj)
+
+        if isinstance(obj, dict):
+            d = obj
+        elif hasattr(obj, "__dict__"):  # object
+            d = obj.__dict__
+        else:
+            raise TypeError("Unexpected obj type: %s" % type(obj))
+
+        if convert_fields:
+            return tuple([conv(d.get(name)) for name, conv in zip(names, converters)])
+        else:
+            return tuple([d.get(name) for name in names])
+
+    return convert_row
+
+
+def _to_java_type(data_type):
+    """
+    Converts Python type to Java type.
+    """
+    gateway = get_gateway()
+    return gateway.jvm.org.apache.flink.table.api.Types.fromJson(data_type.json())
```

Comment:
Serializing all types to JSON format can reduce the communication of many JVMs, which is a good idea. But I think the other issue is we should add the `fromJson()`, i.e. add a new dependency in java module, and introducing maintenance costs on the Java side, which I left the text in Java section.  :)

Reply:
Good idea. I have removed method fromJson and updated method `_to_java_type` to create Java TypeInformation using Py4J.